v4/export/writer_util.go

```diff
@@ -114,6 +120,64 @@ func (b *writerPipe) ShouldSwitchStatement() bool {
 		(b.statementSizeLimit != UnspecifiedSize && b.currentStatementSize >= b.statementSizeLimit)
 }
 
+// SpeedLimiter used for control speed of dump data
+type SpeedLimiter struct {
+	count  int
+	limit  uint64
+	ticker *time.Ticker
+	size   uint64
+	lock   sync.RWMutex
+	tctx   *tcontext.Context
+}
+
+// CheckSpeed check current speed of dump, if already over, sleep the rest time
+func (sl *SpeedLimiter) CheckSpeed(newSize uint64) uint64 {
+	sl.lock.Lock()
+	sl.size += newSize
+	sl.lock.Unlock()
+
+	if sl.size >= sl.limit {
+		sleepTime := 100 - sl.count
+		t := time.NewTicker(time.Millisecond * 10)
+		sl.tctx.L().Debug("We got speed thread ", zap.Uint64("thread", sl.limit), zap.Uint64("used", sl.size), zap.Int("only used time in ms", sl.count*10))
+		for i := 0; i < sleepTime; i++ {
+			<-t.C
+		}
+		t.Stop()
+	}
+	return sl.limit - sl.size
+}
+
+// IntervalCheck check interval auto
+func (sl *SpeedLimiter) IntervalCheck() {
+	go func() {
+		for {
+			<-sl.ticker.C
+			if sl.count >= 100 {
+				sl.lock.Lock()
+				sl.count = 0
+				sl.size = 0
+				sl.lock.Unlock()
+
+				sl.tctx.L().Debug("We resize speed thread.")
+			} else {
+				sl.count++
+			}
+		}
```

Comment:
Please add a context and a wait group to make sure this function can end this loop when dumpling finishes dumping.

Reply:
I adjusted the whole thing.